### PR TITLE
Impl: JWT claim authorization

### PR DESCRIPTION
**Pull request by zhaohuabing**
Implementation for:  https://github.com/envoyproxy/gateway/pull/4009
Related issue: https://github.com/envoyproxy/gateway/issues/3367

This PR also changes the default value of`PayloadInMetada` of the JWT filter from `JwtProvider.Issuer` to `JwtProvider.Name` as `JwtProvider.Issuer` is optional and we need the metadata in the authorization.

### CODE REVIEW

File: examples/kubernetes/http-routing.yaml [hunk 1]
```diff
@@ -70,7 +70,7 @@ spec:
             requests:
               cpu: 10m
 ---
-apiVersion: gateway.networking.k8s.io/v1
+apiVersion: gate  way.networking.k8s.io/v1
```
Comment (zetaab) [inline examples/kubernetes/http-routing.yaml:73]: not needed change?

**Review by zetaab — COMMENTED** [review]
one nit, but I am also thinking that which JWT this is now going to use? There is use-case for both `access_token` and `id_token`. Is there way to specify which one it should use?

**Comment by zhaohuabing** [comment]
> one nit, but I am also thinking that which JWT this is now going to use? There is use-case for both `access_token` and `id_token`. Is there way to specify which one it should use?

It's up to you. A JWT provider must be defined in the JWT configuration within the same `SecurityPolicy`.

If we want to use the ID Token retreived from the OIDC, we need to modify the OAuth2 filter to allow it to add the ID Token in the headers.

**Comment by zetaab** [comment]
I can get this working with idtoken by following configuration (it verifies sub AND that I have some single group that is needed... just for testing)

```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: demo-api-jwt
  namespace: demo
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: demo-api
  jwt:
    providers:
    - name: cognito
      extractFrom:
        cookies:
        - IdToken
      claimToHeaders:
      - claim: sub
        header: x-sub
      - claim: custom:jobtitle
        header: x-jobtitle
      - claim: custom:department
        header: x-department
      - claim: email
        header: x-email
      - claim: cognito:groups
        header: x-groups
      - claim: name
        header: x-name
      remoteJWKS:
        uri: https://cognito-idp.eu-central-1.amazonaws.com/xxx/.well-known/jwks.json
  authorization:
    defaultAction: Deny
    rules:
    - action: Allow
      principal:
        jwt:
          provider: cognito
          claims:
          - name: sub
            values: [ca28333f-177f-4b7a-90c3-snip]
          - name: cognito:groups
            valueType: StringArray
            values: [AAD_FOOBAR]
```

great job @zhaohuabing

**Comment by zhaohuabing** [comment]
> I can get this working with idtoken by following configuration (it verifies sub AND that I have some single group that is needed... just for testing)
@zetaab Thanks for verifying it!

File: internal/xds/translator/authorization.go [hunk 2]
```diff
@@ -262,6 +278,243 @@ func (*rbac) patchRoute(route *routev3.Route, irRoute *ir.HTTPRoute) error {
...
+				{
+					Segment: &networkinput.DynamicMetadataInput_PathSegment_Key{
+						Key: "scope",
+					},
+				},
+			},
+		}
+
+		// The scope is a space-delimited string, so we use a regex matcher to match the scope.
+		pattern := fmt.Sprintf(`(^|.*\s)%s(\s.*|$)`, scope)
```
Comment (arkodg) [inline internal/xds/translator/authorization.go:355]: @denniskniep had presented a nice suggestion earlier of using https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto#extensions-filters-http-jwt-authn-v3-jwtprovider-normalizepayload, that may make this faster

File: api/v1alpha1/authorization_types.go [hunk 3]
```diff
@@ -76,12 +81,22 @@ type Principal struct {
 //
 // +kubebuilder:validation:XValidation:rule="(has(self.claims) || has(self.scopes))",message="at least one of claims or scopes must be specified"
 type JWTPrincipal struct {
+	// Provider is the name of the JWT provider that used to verify the JWT token.
+	// The provider must be configured in the JWT authentication.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Provider string `json:"provider"`
```
Comment (arkodg) [inline api/v1alpha1/authorization_types.go:89]: can we do without this or is this needed ?

File: test/e2e/testdata/authorization-jwt.yaml [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  targetRefs:
+  - group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: http-with-authorization-jwt-combined
+  jwt:
+    providers:
+    - name: example
+      issuer: https://foo.bar.com
+      remoteJWKS:
+        uri: https://www.zhaohuabing.com/download/jwks.json
```
Comment (arkodg) [inline test/e2e/testdata/authorization-jwt.yaml:123]: can we use GH here instead

File: api/v1alpha1/authorization_types.go (hunk 3 above)
Comment (zhaohuabing) [reply api/v1alpha1/authorization_types.go:89]: It's a must as the JWT auth filter stores the extracted JWT claims under its namespace, and we need this to correlate the JWT principal with the JWT claims in the metadata.

```yaml
   envoy.filters.http.jwt_authn:
      jwt_provider_name:
         scopes: "read write"
         name: "John"
         sub: "John"
```

File: test/e2e/testdata/authorization-jwt.yaml (hunk 4 above)
Comment (zhaohuabing) [reply test/e2e/testdata/authorization-jwt.yaml:123]: Sure, I planned to change it to the EG Github one after this PR https://github.com/envoyproxy/gateway/pull/4200, but forgot it. :-)

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zirain — APPROVED** [review]

**Pull request merged**